src/main/java/org/folio/rest/impl/EholdingsTitlesImpl.java

```diff
@@ -0,0 +1,110 @@
+  private TitleParametersValidator parametersValidator;
+  public EholdingsTitlesImpl() {
+    this(
+      new RMAPIConfigurationServiceCache(
+        new RMAPIConfigurationServiceImpl(new ConfigurationClientProvider())),
+      new HeaderValidator(),
+      new TitleParametersValidator(),
+      new TitleConverter());
+  }
+
+  public EholdingsTitlesImpl(RMAPIConfigurationService configurationService,
+                                HeaderValidator headerValidator, TitleParametersValidator parametersValidator,
+                             TitleConverter converter) {
+    this.configurationService = configurationService;
+    this.headerValidator = headerValidator;
+    this.converter = converter;
+    this.parametersValidator = parametersValidator;
+  }
+
+  @Override
+  @Validate
+  @HandleValidationErrors
+  public void getEholdingsTitles(String filterSelected, String filterType, String filterName, String filterIsxn, String filterSubject,
+                                 String filterPublisher, String sort, int page, int count, Map<String, String> okapiHeaders,
+                                 Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    headerValidator.validate(okapiHeaders);
+    parametersValidator.validate(filterSelected, filterType, filterName, filterIsxn, filterSubject, filterPublisher, sort);
+    Sort nameSort = sort != null ? Sort.valueOf(sort.toUpperCase()) : null;
+    CompletableFuture.completedFuture(null)
+      .thenCompose(o -> configurationService.retrieveConfiguration(new OkapiData(okapiHeaders)))
+      .thenCompose(rmapiConfiguration -> {
+        RMAPIService rmapiService = new RMAPIService(rmapiConfiguration.getCustomerId(), rmapiConfiguration.getAPIKey(),
+          rmapiConfiguration.getUrl(), vertxContext.owner());
+        return rmapiService.retrieveTitles(filterSelected, filterType, filterName, filterIsxn, filterSubject,
+          filterPublisher, nameSort, page, count);
+      })
+      .thenAccept(titles ->
+        asyncResultHandler.handle(Future.succeededFuture(GetEholdingsTitlesResponse
+          .respond200WithApplicationVndApiJson(converter.convert(titles)))))
+      .exceptionally(e -> {
+        logger.error(GET_TITLES_ERROR_MESSAGE, e);
+        if(e.getCause() instanceof RMAPIServiceException){
+          RMAPIServiceException rmApiException = (RMAPIServiceException)e.getCause();
+          asyncResultHandler.handle(Future.succeededFuture(Response
+            .status(rmApiException.getRMAPICode())
+            .header("Content-Type", "application/vnd.api+json")
+            .entity(ErrorUtil.createErrorFromRMAPIResponse(rmApiException))
+            .build()));
+        }
+        else {
+          asyncResultHandler.handle(Future.succeededFuture(Response
+            .status(500)
+            .header("Content-Type", "application/vnd.api+json")
+            .entity(ErrorUtil.createError(e.getCause().getMessage()))
+            .build()));
```

Comment:
we could probably extract the code in "if" and "else" into a method and just pass the status code as param to that method since code is duplicated?

Reply:
Now that I look at the code, I realize that the type of error response that we are using from ErrorUtil differs, one uses `ErrorUtil.createErrorFromRMAPIResponse` and the other uses `ErrorUtil.createError` with different exceptions being passed in. So, I think its fine the way it is.